### DISCUSSION TITLE
5V I2C Bus

### DISCUSSION THREAD

**Discussion by AlecHub** [FIRST-Tech-Challenge/FtcRobotController#321] [Ideas]
When connected to a I2C port of a REV Hub, the [REV Logic Level Converter](https://www.revrobotics.com/rev-31-1389/) yields a **5V I2C Bus**.  

The **5V I2C Bus** works flawlessly with Modern Robotics I2C devices:

![LL2_25](https://user-images.githubusercontent.com/15099062/173194696-b08d65bc-86c2-45b4-957e-8439b273557e.jpg)

Kind of like a home-grown CDIM (just the I2C part of the CDIM that is).

**Having to use a separate logic level converter for each 5V I2C device is a barrier to using 5V I2C devices.**  

We would like to submit a rule change request that would allow any number of 5V I2C devices to be connected to the logic level converter (up to the current limit of the converter).

Are there any technical issues with this?

**Comment by Windwoes** [comment]
I can't immediately think of any technical issues, as long as current draw and capacitance is kept under control.

**Reply by AlecHub** [reply]
> I can't immediately think of any technical issues, as long as current draw and capacitance is kept under control.

Thanks @Windwoes.  I believe the max current of a I2C port is 500 mA.  Don't know about capacitance. 

I just found out that the application note  ([AN10441](https://assets.nexperia.com/documents/application-note/AN10441.pdf)) of the chip in the REV Logic Level Converter says the chip is designed **_"to interconnect two sections of an I2C-bus system, with each section having a different supply voltage and different logic levels":_**

![Screenshot 2022-06-23 134559](https://user-images.githubusercontent.com/15099062/175362127-b1835625-436d-4b9e-a7df-948b9eff08da.png)

So the chip is designed to allow for more than one device on each voltage section, and to interconnect all devices in all sections.

**Comment by texasdiaz** [comment]
@AlecHub Let's talk about the REV Logic Level Converter (REV-31-1389). The design intent of the REV Logic Level Converter was to be 1:1 with a sensor. I'm not exactly sure what the maximum current is for an individual I2C bus on the Control Hub/Expansion Hub, but I know that the regulator within the Logic Level Converter can only provide up to 100mA of current at 5V; however, with the included power LED and some quiescent current, that limit is probably closer to around 90mA. Theoretically multiple devices can be connected as long as their total current doesn't exceed ~90mA. There's a HUGE caveat to that, however: depending on the sensor, and any connected pull-up resistors, there is the possibility of degradation of the waveform edges or glitches from instantaneous current demands when switching the I2C data and clock lines. 

So while I don't _**fundamentally**_ have an issue with connecting multiple sensors to a single converter, any team attempting this would need to proceed with LOTS of testing to make sure they don't exceed the current supply limit and can operate glitch-free. But is this really another shoot-yourself-in-the-foot moment that we want to encourage?

**Reply by AlecHub** [reply]
Hi Danny, thanks for the feedback and the info.  It's good to see a working discussion forum.

The documentation on the [REV Logic Level Converter](https://docs.revrobotics.com/duo-control/sensors/5v-sensors#logic-level-converter) mentions the use of the 10441 chip for logic level shifting, but makes no mention of the 3.3V to 5V voltage regulator.  It's good to know that the current limit of the regulator is 100 mA at 5V.  Thank you for the info.

According to the MR specs, the **_maximum_** current draw of Modern Robotics' I2C devices is 20 mA.  So I would recommend no more than 4 Modern Robotics I2C devices per REV Logic Level Converter. 

Alternatively, you could supply power to the 5V I2C devices via the [5V Power Port of a REV Hub](https://docs.revrobotics.com/duo-control/control-system-overview/expansion-hub-basics#+5v-power-port-specifications) instead of the REV Logic Level Converter per the following circuit:

![I2CPM_11](https://user-images.githubusercontent.com/15099062/175781939-2fe7e2c6-adc9-4ded-b2ed-e33029af47b8.jpg)

The PWM cable in the circuit is connected a 5V Power Port of a REV Hub, which is rated at 2000 mA (enough to power 100 MR I2C devices).  I've tested with 7 MR devices and the current draw is between 110 and 135 mA, depending on the color of the Color Beacons.

**Reply by AlecHub** [reply]
> The documentation on the [REV Logic Level Converter](https://docs.revrobotics.com/duo-control/sensors/5v-sensors#logic-level-converter) mentions the use of the 10441 chip for logic level shifting ...

Apparently [AN10441](https://assets.nexperia.com/documents/application-note/AN10441.pdf) is not a chip; it is a technique.  I am not an EE, _I just play one on the forums_ ;)
